File: 1110. Delete Nodes And Return Forest/1110. Delete Nodes And Return Forest.py
class Solution:
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        to_delete_set = set(to_delete)
        remaining_trees = []

        def helper(node, is_root):
            if not node:
                return None
            should_delete = node.val in to_delete_set
            node.left = helper(node.left, should_delete)
            node.right = helper(node.right, should_delete)
            if not should_delete and is_root:
                remaining_trees.append(node)
            return None if should_delete else node
        
        helper(root, True)
        return remaining_trees

File: 1110. Delete Nodes And Return Forest/1110. Delete Nodes And Return Forest.cpp
class Solution {
public:
    const static int N = 1001;
    bitset<N> rmv = 0;
    vector<TreeNode*> ans;

    void dfs(TreeNode* root, TreeNode* parent, bool isLeft) {
        if (!root) return;
        if (rmv[root->val]) {
            if (parent) {
                if (isLeft) parent->left = NULL;
                else parent->right = NULL;
            }
            dfs(root->left, NULL, 1);
            dfs(root->right, NULL, 0);
            delete root;  
        } 
        else {
            if (!parent) 
                ans.push_back(root);

            dfs(root->left, root, 1);
            dfs(root->right, root, 0);
        }
    }

    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        for (int x : to_delete) 
            rmv[x] = 1;
        dfs(root, NULL, 1);
        return ans;
    }
};